mshadow/random.h

```diff
@@ -68,47 +69,166 @@ class Random<cpu, DType> {
    */
   inline void set_stream(Stream<cpu> *stream) {
   }
+
+// These samplers are only avail in C++11.
+#if MSHADOW_IN_CXX11
+
   /*!
-   * \brief generate data from uniform [a,b)
+   * \brief get some random integer
+   * \return integer as unsigned
+   */
+  inline unsigned GetRandInt() {
+    return rnd_engine_();
+  }
+
+  /*!
+   * \brief generate data from a distribution
    * \param dst destination
-   * \param a lower bound of uniform
-   * \param b upper bound of uniform
    * \tparam dim dimension of tensor
+   * \param sampler sampler of the distribution
    */
-  template<int dim>
-  inline void SampleUniform(Tensor<cpu, dim, DType> *dst,
-                            DType a = 0.0f, DType b = 1.0f) {
+  template<int dim, class Sampler>
+  inline void SampleDistribution(Tensor<cpu, dim, DType> *dst, Sampler sampler) {
     if (dst->CheckContiguous()) {
-      this->GenUniform(dst->dptr_, dst->shape_.Size(), a, b);
+      std::generate_n(dst->dptr_, dst->shape_.Size(), sampler);
     } else {
       Tensor<cpu, 2, DType> mat = dst->FlatTo2D();
       for (index_t i = 0; i < mat.size(0); ++i) {
-        this->GenUniform(mat[i].dptr_, mat.size(1), a, b);
+        std::generate_n(mat[i].dptr_, mat.size(1), sampler);
       }
     }
   }
+
+  /*!
+   * \brief generate data from uniform [a,b)
+   * \param dst destination
+   * \param a lower bound of uniform
+   * \param b upper bound of uniform
+   * \tparam dim dimension of tensor
+   */
+  template<int dim, typename PType>
+  inline void SampleUniform(Tensor<cpu, dim, DType> *dst,
+                            PType a , PType b ) {
```

Comment:
Corrected this and retained default values for uniform/gaussian. Pushed it again. Will wait to deal with your remarks concerning the bigger mxnet part after this is pulled in as only then I can get a clean compile & test.